### PR TITLE
[GUFA] Infer a RefEq value of 0 when possible

### PR DESCRIPTION
**Pull request by kripken**
If the PossibleContents for the two sides have no possible intersection then the
result must be 0.

### CODE REVIEW

**Comment by kripken** [comment]
This helps in several places in j2wasm output.

**Review by tlively — APPROVED** [review]

File: src/ir/possible-contents.cpp [hunk 1]
```diff
@@ -992,6 +1026,12 @@ struct Flower {
     return locations[index].contents;
   }
 
+  // Helper for the common case of an expression location that is not a
+  // multivalue.
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/ir/possible-contents.cpp:1033]: Added, thanks.

File: src/ir/possible-contents.cpp [hunk 2]
```diff
@@ -132,6 +132,45 @@ void PossibleContents::combine(const PossibleContents& other) {
...
+    // One is the set of all things, so definitely something can intersect since
+    // we've ruled out an empty set for both.
+    return true;
+  }
+
+  auto aType = a.getType();
+  auto bType = b.getType();
+
+  if (aType.isNullable() && bType.isNullable()) {
+    // Null is possible on both sides, regardless of the types.
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/ir/possible-contents.cpp]: Good point. I added a TODO now. Not sure assertions are worthwhile, as they'd prevent the tests from running.

Or I guess I could implement the full check now? What is the proper way to check if types are in separate hierarchies - no lub, as you said?

[the remark this answers is not in the document]
Comment (tlively) [reply src/ir/possible-contents.cpp]: Yeah, checking for no LUB is the simplest option right now.

[the remark this answers is not in the document]
Comment (kripken) [reply src/ir/possible-contents.cpp]: I see, thanks.

Thinking on this more, I think we should leave it as a TODO for now. Making it more optimal would add complexity and runtime overhead, and we gain nothing from it atm.

**Review by tlively — APPROVED** [review]

**Comment by tlively** [comment]
I believe this is similar to the ref.eq change where handling the monotonicity properly should allow us to propagate the zero or one while still flowing values around.

**Comment by kripken** [comment]
Yes, I think we could probably keep these in the main flow, with more logic to handle the corner case that was wrong with monotonicity (if one of the inputs is None then we should not even query if there is an intersection). But I think it's simpler this way to keep it out of the flow, and it can't help the rest of the flow to do any more than it can atm. (But in the flow it might reduce the number of iterations.)

**Pull request merged**